---
title: "Load - Daims Nitrospira Comammox VetMed assembly"
author: "Mads Albertsen"
date: "November 24, 2015"
output: html_document
---

## Load packages

In case you havn't installed all the needed packages, they can be installed via e.g. `install_github("MadsAlbertsen/mmgenome/mmgenome")`.
```{r Load_packages, message=F,warning=FALSE}
library("mmgenome")
options(scipen = 8)
```

## Load data

The assembly in fasta format is loaded using the Biostrings package.
```{r Load_assembly}
assembly <- readDNAStringSet("data/assembly.fa", format = "fasta")
```

The coverage data is loaded. Note that each coverage dataset must have the scaffold name in the first column and the coverage in the second coloumn. The name of each coverage dataset reflect when the data was sampled.
```{r Load_coverage}
VM23 <- read.table("data/VM23.csv", header = T, sep = ",")[,c("Reference.sequence", "Average.coverage")]
VMPS <- read.table("data/VMPS.csv", header = T, sep = ",")[,c("Reference.sequence", "Average.coverage")]
```

The overview of the number of essential genes in each scaffold is loaded.
```{r Load_essential}
ess <- read.table("data/essential.txt", header = T, sep = " ")
```

Load basic taxonomy of the scaffolds. Only scaffolds with essential genes are taxonomic classified.
```{r Load_tax}
tax <- read.table("data/tax.txt", header = T, sep = "\t")
```

Load information on which scaffolds that are connected by either paired-end (PE) or mate-pair (MP) connections.
```{r Load_network}
pe <- read.table("data/network_pe.txt", header = T, sep = "\t")
```

Additional: 16S rRNA infomration.
```{r Load_Additional_16S}
rRNA16S <- read.table("data/16S.txt", header = T, sep = "\t")
```

Additional: ESOM
```{r Load_Additional_esom}
esom_names <- read.delim("data/esom.names", skip = 1, header = F)
esom_bm <- read.delim("data/esom.bm", skip = 2, header = F)
esom_umx <- read.delim("data/esom.umx", skip = 1, header = F)

esom_data <- mmformat_esom(names = esom_names, bm = esom_bm, umx = esom_umx)

esom_map <- esom_data$esom_map
esom <- esom_data$esom_coord
```

## Merge data

The loaded data is merged to a single object. The assembly is used to extract scaffold length, GC contente and tetranucleotide frequencies, which are used for a simple PCA. The `tax.expand` parameter converts a phylum level taxonomic classification to the underlying classes instead. The `tax.freq` parameter removes any classification with less than 50 entries in the complete dataset. The `other` parameter can be used to load any additional datasets.
```{r Merge_data}
d <- mmload(assembly = assembly, 
            coverage = c("VM23", "VMPS"),
            essential = ess,           
            tax = tax,
            tax.expand = "Proteobacteria",
            tax.freq = 25,
            pca = T,
            other = c("rRNA16S", "esom")
           )
```

## Save data
Remove temporary data and save the generated data for easy loading.
```{r Save_data}
rm(list = c("ess", "VM23", "VMPS", "rRNA16S", "tax", "esom_bm", "esom_names", "esom_umx", "esom", "esom_data"))
save.image(file="Daims_VetMed.RData")
```