---
title: "Analysis - Overview of the GWW assembly"
author: "Mads Albertsen"
date: "November 24, 2015"
output: html_document
---
## Introduction
This report documents the GWW dataset, where mulitple Comammox Nitrospira genome bins were extracted. See [Daims et al., 2015: Complete Nitrification by Nitrospira Bacteria](https://doi.org/10.1038/nature16461) for further details.

## Load the mmgenome package
```{r Load_package, message=FALSE, warning=FALSE, results='hide'}
library("mmgenome")
```

## Import data
The Rmarkdown file [Load_data.Rmd](http://madsalbertsen.github.io/mmgenome/Daims_GWW_Load_data.html) describes the loading of the data and can be imported using the `mmimport` function. However, the preprocessed data can also be downloaded directly from figshare: [Daims_VetMed](). Hence, here we import the prepocessed data from figshare instead.
```{r Load_data, message=FALSE, warning=FALSE, results='hide'}
load("Daims_GWW.RData")
```

## Data overview
The object `d` contains information on scaffolds and essential genes within the scaffolds. For each scaffold the dataset contains the following information: The columns `HPD` and `HPF1` contain the coverage information from 4 different samples; `PC1`, `PC2` and `PC3` contain coordinates of the three first principal components from a PCA analysis on tetranucleotide frequencies; `essential` contain information taxonomic information for each scaffold based on classification on essential genes; `rRNA16S` contain taxonomic information on scaffolds that have an associated 16S rRNA gene.
```{r overview_names}
colnames(d$scaffolds)
```

```{r overview_stats}
mmstats(d, ncov = 2)
```

## Extract nitrospira scaffolds
Scaffolds with essential genes classified to Nitrospira are extracted for a cleaner visualisation later.
```{r nit}
nit <- subset(d$scaffolds, essential == "Nitrospirae")
```

## Overview plot
The assembly is decent even though it is from a full-scale sample with a high degree of micro-diversity.
```{r overview, fig.width=12, fig.height=10, fig.align='center', warning=FALSE, message=FALSE, echo=FALSE}
mmplot(data = d, 
       x = "HPD", 
       y = "HPF1", 
       log.x = T, 
       log.y = T, 
       color = "essential",
       minlength = 5000) +
  scale_x_log10(limits=c(0.1, 500), 
                breaks = c(0.1,1,10,100)) +
  scale_y_log10(limits=c(0.1, 500), 
                breaks = c(0.1,1,10,100)) +
  scale_size_area(breaks = c(10000, 50000,  100000, 500000, 1000000), 
                  max_size = 15, 
                  labels = c(10, 50, 100, 500, 1000), 
                  name = "Scaffold Length (Kbp)") +
  scale_color_discrete(name = "Taxonomy") +
  xlab("Coverage (GWW HP_D)") +
  ylab("Coverage (GWW HP_F1)") +
  theme(panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(color = "black"),
        axis.ticks = element_line(color = "black"),
        axis.text = element_text(color = "black"),
        legend.key = element_blank())
```

## Overview plot - Nitrospira
Here I've highlighted the scaffolds that contains essential genes from Nitrospira. There is a large number of abundant Nitrospira! Some assembles very nicely, while others are affected by micro-diversity.
```{r overview_nit, fig.width=12, fig.height=10, fig.align='center', warning=FALSE, message=FALSE, echo=FALSE}
mmplot(data = d, 
       x = "HPD", 
       y = "HPF1", 
       log.x = T, 
       log.y = T, 
       color = "essential", 
       minlength = 5000) +
  geom_point(data = nit, color = "red") +
  scale_x_log10(limits=c(0.1, 500), 
                breaks = c(0.1,1,10,100)) +
  scale_y_log10(limits=c(0.1, 500), 
                breaks = c(0.1,1,10,100)) +
  scale_size_area(breaks = c(10000, 50000,  100000, 500000, 1000000), 
                  max_size = 15, labels = c(10, 50, 100, 500, 1000), 
                  name = "Scaffold Length (Kbp)") +
  scale_color_discrete(name = "Taxonomy") +
  xlab("Coverage (GWW HP_D)") +
  ylab("Coverage (GWW HP_F1)") +
  theme(panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(color = "black"),
        axis.ticks = element_line(color = "black"),
        axis.text = element_text(color = "black"),
        legend.key = element_blank())
```