---
title: "Example: Genome Extraction"
author: "Mads Albertsen"
date: "Sunday, August 24, 2014"
output: html_document
---

## Load the mmgenome package
In case you havn't installed the [mmgenome package](http://madsalbertsen.github.io/mmgenome/), it can be installed using the `devtools` package through github: `devtools::install_github("MadsAlbertsen/mmgenome/mmgenome")`.
```{r Load_package, message=FALSE, warning=FALSE, results='hide'}
library("mmgenome")
```

## Import data
The Rmarkdown file `Load_data.Rmd` describes the data that needs to be generated and loaded. The data is loaded using the `mmimport` function. The data loading and genome extraction is split to enable cleaner workflows.
```{r Load_data, message=FALSE, warning=FALSE, results='hide'}
#mmimport(file = "Load_data.Rmd")
```

However, in this example we load the attached example dataset.
```{r Load_example}
data(rocco)
```

## Data overview
The object `d` contains information on scaffolds and essential genes within the scaffolds. The `rocco` dataset contains the following information on each scaffold.
```{r overview_names}
colnames(d$scaffolds)
```

The basic statistics of the full dataset can be summarised using the `mmstats` function. 
```{r overview_stats}
mmstats(d, ncov = 4)
```

## Initial scaffold extraction
The `mmplot` function is used to generate a differential coverage plot. The plot is stored in the object `p` and then plotted using `p`. `mmplot_locator` is used to interactively define a subspace to extract on the plot. For complete reproducibility the coordinates of the subspace is also stored manually (the coordinates are written to the console).  `mmplot_selection` is used plot the subspace in the final markdown report. To use `mmplot_locator`, simply plot the data using `p` and then run `mmplot_locator(p)` to interactively define the subspace, press `finish` when done.

Here choose an Actinobacteria that seems like an easy target.
```{r zoomA, fig.width=12, fig.height=10, fig.align='center'}
p <- mmplot(data = d, 
            x = "C13.11.25", 
            y = "C14.01.09", 
            log.x = T, 
            log.y = T, 
            color = "essential",
            minlength = 3000)

#p
#sel <- mmplot_locator(p)

sel <- data.frame(C13.11.25  =  c(7.2, 16.2, 25.2, 23.3, 10.1),
                  C14.01.09  =  c(47, 77, 52.8, 29.5, 22.1))

mmplot_selection(p, sel)
```

The scaffolds included in the defined subspace are extracted using the `mmextract` function. Note: all scaffolds in the subspace are extracted even though only scaffolds over 3000 bp were used for plotting in the example above. `mmextract` also contains a `minlength` parameter if needed.
```{r extractA}
dA <- mmextract(d, sel)
```

The `mmstats` function applies to any extracted object. Hence, it can be used directly on the subset.
```{r statsA}
mmstats(dA, ncov = 4)
```

To get a more clear look on what was in the initial subset we simply plot it using the `mmplot` function. Note that the `dA` object is used as input.
```{r replotA, fig.width=12, fig.height=10, fig.align='center'}
mmplot(data = dA, 
       x = "C13.11.25", 
       y = "C14.01.09", 
       log.x = T, 
       log.y = T,
       color = "essential")
```

## Finding the relevant data for next subset
Although it seems like we would be able to extract the Actinobacteria using the intial two coverage datasets, we'll take a look at the additional data we have available. 

The `mmplot_pairs` function can be used to plot a number of variables against each other, to get a quick overview of which might be used to seperate the genomes of interest. The scaffolds are colored with the PhyloPythia+ taxonomic classification.
```{r pairsA, fig.width=12, fig.height=10, fig.align='center'}
mmplot_pairs(data = dA,
             variables = c("C13.11.14","C13.11.25","C13.12.03", "C14.01.09", "gc", "PC2"), 
             log = c("C13.11.14","C13.11.25","C13.12.03", "C14.01.09"),
             color = "pps_phylum",
             textsize = 5
             )
```

Using the `C13.11.14` and `C13.12.03` coverage profiles might enable us to remove the leftover contaminants. Hence, they are used for a new subspace extraction.

## Subspace extraction 2
We use the same prodcedure as with the initial subset. However, now the `dA` subset is used as input data.
```{r zoomB, fig.width=12, fig.height=10, fig.align='center'}
p <- mmplot(data = dA, 
            x = "C13.12.03", 
            y = "C13.11.14", 
            log.x = T, 
            log.y = T, 
            color = "pps_phylum")

#p
#sel <- mmplot_locator(p)

sel <- data.frame(C13.12.03  =  c(3.48, 4.95, 6.97, 13.6, 15.7, 9.68, 4.48),
                  C13.11.14  =  c(0.407, 1.72, 2.92, 1.45, 0.264, 0.17, 0.163))

mmplot_selection(p, sel)
```

The scaffolds in the subspace are extracted and stored in the object `dB`. Note that `dA` is now used as the input dataset.
```{r extractB}
dB <- mmextract(dA, sel)
```

... and `mmstats` is used to check the extracted bin.
```{r statsB}
mmstats(dB, ncov = 4)
```

## Using paried-end connections
Until now we have just used the coverage profiles to extract scaffolds related to our genome of interest. However, some scaffolds might be present in many copies (repeats) and hence have a much higher coverage than the rest of the genome. In addition, some scaffolds will by chance have a slightly different coverage profile than the rest of the genome and thereby also been missed.

The function `mmplot_network` can be used to generate a network plot of scaffolds connected by paired-end reads. We start by plotting the scaffolds we have in our current subset.
```{r networkB, fig.width=10, fig.height=10, fig.align='center'}
mmplot_network(data = dB, network = pe, nconnections = 1, color = "essential")
```

Most scaffolds seem to be nicely connected and we should be able to improve the assembly substantial in the refinement stage.

To include repeats and other missed scaffolds we simply extract all scaffolds that are directly connected by paired-end reads to our current subset `dB` using `mmextract_network`. Only scaffolds directly connected to the subset is extracted. 
```{r network_extractB}
dC <- mmextract_network(subset = dB, original = d, network = pe, nconnections = 1, type = "direct")
```

... and then plot the new subset. After the network extraction we have 4 seperate clusters of scaffolds that are connected. This means that we had some scaffolds in the `dB`subset that didn't belong to our target genome. However, we have also included a few correct repeats that belongs to the target Actinobacteria. Compared to the previous network plot it can also be seen that a number of small scaffolds are now connected to the Actinobacteria bin. This is low abundant micro-diversity.
```{r networkC, fig.width=10, fig.height=10, fig.align='center'}
mmplot_network(data = dC, network = pe, nconnections = 1, color = "pps_phylum")
```

## Subspace extraction 3
To exclude the contamination and include the correct repeats we make another subspace extraction. To make it more easy to spot the small scaffolds we remove the length scale from the scaffolds by setting a fixed `point.size` .
```{r zoomC, fig.width=12, fig.height=10, fig.align='center'}
p <- mmplot(data = dC, x = "C14.01.09", y = "C13.11.25", log.x = T, log.y = T, color = "pps_phylum", point.size = 5)

#p
#sel <- mmplot_locator(p)

sel <- data.frame(C14.01.09  =  c(19, 15.5, 69.1, 103, 106, 54.9),
                  C13.11.25  =  c(7.28, 17.2, 163, 165, 21.4, 7.76))

mmplot_selection(p, sel)
```

Now, before we go ahead and extract the scaffolds within the subspace we have to remove the contaminating scaffolds that were present inside our bin. As we are using the same variables for subsetting as before, they will still be there. One way to remove them is simply to identify their scaffold names. 

This can be done by plotting the previous network graph of the current subset `dC` and labelling the scaffolds using the previous subset `dB`. The contaminating scaffolds are `10932`, `29668` and `50917`.
```{r networkC2, fig.width=10, fig.height=10, fig.align='center'}
mmplot_network(data = dC, network = pe, nconnections = 1, color = "pps_phylum", highlight = dB)
```

The `mmextract` function is used to extract the defined subset and at the same time exclude the manually specified scaffolds.
```{r mmextractD}
dD <- mmextract(data = dC, selection = sel, exclude = c("10932", "29668", "50917"))
```

## Final overview
The statistics of the final subset can be seen using `mmstats`.
```{r mmstatsD}
mmstats(dD, ncov = 4)
```

The extracted scaffolds are connected quite nicely by PE reads and it might even be possible to close the genome using the MP data.
```{r networkD, fig.width=10, fig.height=10, fig.align='center'}
mmplot_network(data = dD, network = pe, nconnections = 1, color = "pps_phylum")
```

The extracted subset can also be highlighted in context of the original the dataset. Note that all plots are ggplot objects and hence can be manipulated in almost any way you can imagine. Here, we just adjust limits of the axis. 
```{r highlight, fig.width=12, fig.height=10, fig.align='center'}
mmplot(data = d, 
       x = "C14.01.09", 
       y = "C13.11.25", 
       log.x = T, 
       log.y = T, 
       color = "none", 
       highlight = dD, 
       minlength = 3000) +
  scale_x_log10(limits=c(0.1,300)) +
  scale_y_log10(limits=c(0.05,200))
```

## Evaluate completeness 
We can also compare the number of essential genes to all complete Actinobacteria in NCBI using the `mmref` function. There are a few duplicated "single copy" essential genes. However, the duplicated genes are also seen in some of the complete Actinobacteria. Hint: Try using tax.aggregate = "Genus". 
```{r mmref, fig.width=10, fig.height=4}
mmref(data=dD, tax.level = "Phylum", tax.compare="Actinobacteria")
```

## Export the scaffolds 
Now that we are happy with the genome bin, the scaffolds can be exported to a seperate fasta file using `mmexport`.
```{r mmexport}
mmexport(data=dD, assembly=assembly, file = "Awesome_actinobacteria.fa")
```