Load Data
=======================================

## Load needed packages

In case you havn't installed all the package, it can be installed via `install_github("MadsAlbertsen/mmgenome/mmgenome")` (reguires devtools).
```{r Load_packages, message=F,warning=FALSE}
library("mmgenome")
options(scipen = 8)
```

## Load all data

The data is loaded and a few of the data columns are renamed.
```{r Load_data}
S1 <- read.csv("data/S1.csv", header = T)               
S2 <- read.csv("data/S2.csv", header = T)

gc <- read.delim("data/assembly.gc.tab", header = T)
kmer <- read.delim("data/assembly.kmer.tab", header = T)
ess <- read.table("data/assembly.orfs.hmm.id.txt", header = F)
cons.tax <- read.delim("data/tax.txt", header = T)

a16S<-read.delim("data/rRNA/16S.csv", header = T, sep = ";")
a23S<-read.delim("data/rRNA/23S.csv", header = T, sep = ";")

assembly <- readDNAStringSet("data/assembly.fa", format = "fasta")

network <- read.delim("data/network.txt", header = T)

colnames(kmer)[1] = "scaffold"
colnames(ess) = c("scaffold","orf","hmm.id")
```

Merge coverage, gc and length data on all scaffolds into a single data frame `d`.
```{r Merge_d1}
d <- cbind.data.frame(gc$contig,                       
                      gc$gc, 
                      S1$Reference.length,
                      S1$Average.coverage, 
                      S2$Average.coverage)
colnames(d) = c("scaffold", "gc", "length", "S1", "S2")
```

Add PCA analysis of tetranucleotide frequencies to `d`.
```{r Merge_PCA}
rda <- rda(kmer[,2:ncol(kmer)])
d<-cbind(d,scores(rda,choices=1:3)$sites)
```

Clean up the consensus taxonomic assignments from essential single copy genes and add them to `d`.
```{r Merge_d_tax_ess}
tax <- clean_tax(cons.tax, occurence=20, expand="Proteobacteria", clean = T)
d <- merge(d,tax[,c(1,2)], by = "scaffold", all = T)
```

Add 16S and 23S rRNA assignments to `d`. The classification is taken directly from arb-silva's online classifier.
```{r Merge_rRNA}
t16S <- import_rrna(data=a16S, type="16S")
t23S <- import_rrna(data=a23S, type="23S")
d <- merge(x = d, y = t16S , by = "scaffold", all = T)
d <- merge(x = d, y = t23S , by = "scaffold", all = T)
```

Store the entire network graph of paired-end connections in the object `g` and attach information from the scaffolds. We extract all clusters from the paired-end graph and add them to `d`. First we find all discrete clusters, i.e. scaffolds that are connected by paired-end reads. Then we add the cluster number and cluster size to `d`.
```{r Network_generation}
g <- import_network(network=network, data=d, nconnections = 10)
d <- merge_network(graph = g, data = d)
```

Merge all data on essential genes into a single data frame `e`.
```{r Merge_e}
e <- merge(ess, d, by = "scaffold", all.x = T)
```

Custom color palette for gc coloring.
```{r Color_palettes}
rgb.p<- colorRampPalette(c('red','green','blue'))
gc.trans<-adjustcolor(rgb.p(max(d$gc)-min(d$gc)),alpha.f=0.3)
```