<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools for extracting individual genomes from metagenomes • mmgenome</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">mmgenome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="articles/mmgenome.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
<li>
  <a href="articles/assembly.html">Assembly guides</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MadsAlbertsen/mmgenome">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="mmgenome" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#mmgenome" class="anchor"></a>mmgenome</h1></div>
<p>mmgenome: Tools for extracting genomes from metagenomes</p>
<div id="changelog" class="section level2">
<h2 class="hasAnchor">
<a href="#changelog" class="anchor"></a>Changelog</h2>
<div id="section" class="section level3">
<h3 class="hasAnchor">
<a href="#section" class="anchor"></a>0.7.1</h3>
<div id="enhancements" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmstats: Now icludes an option to compare the coverage as a fraction of the original dataset (Kirk3gaard).</li>
</ul>
</div>
</div>
<div id="section-1" class="section level3">
<h3 class="hasAnchor">
<a href="#section-1" class="anchor"></a>0.7.0</h3>
<div id="bugfix" class="section level4">
<h4 class="hasAnchor">
<a href="#bugfix" class="anchor"></a>Bugfix</h4>
<ul>
<li>mmplot_locator: Updated to handle the new format of ggplot2 objects (&gt;v. 1.0.0).</li>
</ul>
</div>
</div>
<div id="section-2" class="section level3">
<h3 class="hasAnchor">
<a href="#section-2" class="anchor"></a>0.6.3</h3>
<div id="bugfix-1" class="section level4">
<h4 class="hasAnchor">
<a href="#bugfix-1" class="anchor"></a>Bugfix</h4>
<ul>
<li>mmplot_network: Updated to the newest version of igraph. Igraph changed function and variable names.</li>
</ul>
</div>
</div>
<div id="section-3" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3" class="anchor"></a>0.6.2</h3>
<div id="enhancements-1" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements-1" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmplot: Now includes an option, <code>factor.shape</code> to color factors with solid shapes instead.</li>
<li>mmformat_esom: Formats ESOM data from databionics to a simple dataframe which can be loaded in mmgenome.</li>
<li>mmformat_ppsp: Formats classification data from PhyloPythiaS+ to a simple dataframe which can be loaded in mmgenome.</li>
<li>mmadd: Adds a dataframe to an existing mmgenome object.</li>
</ul>
</div>
</div>
<div id="section-4" class="section level3">
<h3 class="hasAnchor">
<a href="#section-4" class="anchor"></a>0.6.1</h3>
<div id="enhancements-2" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements-2" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmplot: Now includes an option to overlay a ESOM contour plot if ESOM data have been loaded.</li>
</ul>
</div>
</div>
<div id="section-5" class="section level3">
<h3 class="hasAnchor">
<a href="#section-5" class="anchor"></a>0.6.0</h3>
<div id="bugfix-2" class="section level4">
<h4 class="hasAnchor">
<a href="#bugfix-2" class="anchor"></a>Bugfix</h4>
<ul>
<li>mmsubset: The subsetting of essential genes was wrong, which affected the mmstats calculations. Fixed thanks to @russianconcussion.</li>
</ul>
</div>
</div>
<div id="section-6" class="section level3">
<h3 class="hasAnchor">
<a href="#section-6" class="anchor"></a>0.6.0</h3>
<div id="enhancements-3" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements-3" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmplot_network: new option to scale line width of links between scaffolds by a constant (<code>scale.links</code>).</li>
</ul>
</div>
</div>
<div id="section-7" class="section level3">
<h3 class="hasAnchor">
<a href="#section-7" class="anchor"></a>0.5.0</h3>
<div id="enhancements-4" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements-4" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmsubset: new function to subset based on normal subset syntax. E.g. <code>gc &gt; 50</code>.</li>
</ul>
</div>
<div id="bugfixes" class="section level4">
<h4 class="hasAnchor">
<a href="#bugfixes" class="anchor"></a>Bugfixes</h4>
<ul>
<li>mmstats: Now rounding coverages to 2 decimals instead of one (Kirk3gaard).</li>
</ul>
</div>
</div>
<div id="section-8" class="section level3">
<h3 class="hasAnchor">
<a href="#section-8" class="anchor"></a>0.4.3</h3>
<ul>
<li>mmstats: Error in summing scaffold lengths of large datasets due to integer overflow. Handled by converting integer to numeric (skarst).</li>
</ul>
</div>
<div id="section-9" class="section level3">
<h3 class="hasAnchor">
<a href="#section-9" class="anchor"></a>0.4.2</h3>
<ul>
<li>mmplot: log.x and log.y is now false by default in order to function in mmplot_pairs (skarst).</li>
</ul>
</div>
<div id="section-10" class="section level3">
<h3 class="hasAnchor">
<a href="#section-10" class="anchor"></a>0.4.1</h3>
<ul>
<li>mmload: Failed to load data if coverage was missing for some scaffolds. Fixed thanks to Matthew Neave (neavemj).</li>
</ul>
</div>
<div id="section-11" class="section level3">
<h3 class="hasAnchor">
<a href="#section-11" class="anchor"></a>0.4.0</h3>
<ul>
<li>mmplot_network: It’s now possible to to highlight scaffolds.</li>
<li>General: Information on essential genes is no longer required.</li>
</ul>
</div>
<div id="section-12" class="section level3">
<h3 class="hasAnchor">
<a href="#section-12" class="anchor"></a>0.3.1</h3>
<div id="bugfixes-1" class="section level4">
<h4 class="hasAnchor">
<a href="#bugfixes-1" class="anchor"></a>Bugfixes</h4>
<ul>
<li>mmload: Column names of additional datasets did not load as intended.</li>
</ul>
</div>
<div id="enhancements-5" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements-5" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmplot: The variable “point.size” can be used to scale the points by a constant instead of the scaffold length.</li>
<li>mmimport: New simple function to handle import of the data through a Rmarkdown file.</li>
</ul>
</div>
</div>
<div id="section-13" class="section level3">
<h3 class="hasAnchor">
<a href="#section-13" class="anchor"></a>0.3.0</h3>
<div id="enhancements-6" class="section level4">
<h4 class="hasAnchor">
<a href="#enhancements-6" class="anchor"></a>Enhancements</h4>
<ul>
<li>mmplot: It’s now possible to highlight subsets. Either by just suppling the subset dataset or using the names of the scaffolds.</li>
<li>mmplot: “color” support the argument “none” for coloring all scaffolds black.</li>
<li>mmextract: The function now supports to “include” or “exclude” scaffolds based on a list of scaffold names.</li>
</ul>
</div>
</div>
<div id="section-14" class="section level3">
<h3 class="hasAnchor">
<a href="#section-14" class="anchor"></a>0.2.0</h3>
<p>First stable release of the mmgenome package</p>
</div>
<div id="section-15" class="section level3">
<h3 class="hasAnchor">
<a href="#section-15" class="anchor"></a>0.1.0</h3>
<p>Initial release of the mmgenome package</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Latest stable release at <br><a href="https://github.com/MadsAlbertsen/mmgenome/releases/latest/">https://​github.com/​MadsAlbertsen/​mmgenome/​releases/​latest/​</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://www.r-project.org/Licenses/AGPL-3">AGPL-3</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Mads Albertsen <br><small class="roles"> Author, maintainer </small> </li>
<li>Kasper Skytte Andersen <br><small class="roles"> Author </small> </li>
<li>Rasmus Hansen Kirkegaard <br><small class="roles"> Author </small> </li>
<li>Søren Mikael Karst <br><small class="roles"> Author </small> </li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard, Søren Mikael Karst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
